src/main/scala/org/constellation/primitives/ConcurrentTipService.scala

```diff
@@ -0,0 +1,198 @@
+import org.constellation.consensus.RoundManager.{ActiveTipMinHeight, GetActiveMinHeight}
+import org.constellation.consensus.TipData
+import org.constellation.primitives.Schema.{Id, SignedObservationEdge}
+import org.constellation.primitives.concurrency.{SingleLock, SingleRef}
+import org.constellation.util.Metrics
+import org.constellation.{ConstellationContextShift, ConstellationExecutionContext, DAO}
+
+import scala.util.Random
+
+case class TipSoe(soe: Seq[SignedObservationEdge], minHeight: Option[Long])
+case class PulledTips(tipSoe: TipSoe, peers: Map[Id, PeerData])
+case class TipConflictException(cb: CheckpointBlock, conflictingTxs: List[String])
+    extends Exception(
+      s"CB with baseHash: ${cb.baseHash} is conflicting with other tip or its ancestor. With following txs: $conflictingTxs"
+    )
+case class TipThresholdException(cb: CheckpointBlock, limit: Int)
+    extends Exception(
+      s"Unable to add CB with baseHash: ${cb.baseHash} as tip. Current tips limit met: $limit"
+    )
+
+class ConcurrentTipService[F[_]: Concurrent](sizeLimit: Int,
+                                             maxWidth: Int,
+                                             maxTipUsage: Int,
+                                             numFacilitatorPeers: Int,
+                                             minPeerTimeAddedSeconds: Int,
+                                             dao: DAO,
+                                             consensusActor: ActorRef)
+    extends StrictLogging {
+
+  private val conflictingTips: SingleRef[F, Map[String, CheckpointBlock]] = SingleRef(Map.empty)
+  private val tipsRef: SingleRef[F, Map[String, TipData]] = SingleRef(Map.empty)
+  private val semaphore: Semaphore[F] = {
+    implicit val cs: ContextShift[IO] = ConstellationContextShift.edge
+    Semaphore.in[IO, F](1).unsafeRunSync()
+  }
+
+  private def withLock[R](name: String, thunk: F[R]) = new SingleLock[F, R](name, semaphore).use(thunk)
+
+  implicit var shortTimeout: Timeout = Timeout(3, TimeUnit.SECONDS)
+
+  def set(newTips: Map[String, TipData]): F[Unit] =
+    tipsRef.set(newTips)
+
+  def toMap: F[Map[String, TipData]] =
+    tipsRef.get
+
+  def size: F[Int] =
+    tipsRef.get.map(_.size)
+
+  def sizeUnsafe: F[Int] =
+    tipsRef.getUnsafe.map(_.size)
+
+  def get(key: String): F[Option[TipData]] =
+    tipsRef.get.map(_.get(key))
+
+  def getUnsafe(key: String): F[Option[TipData]] =
+    tipsRef.getUnsafe.map(_.get(key))
+
+  def remove(key: String)(implicit metrics: Metrics): F[Unit] =
+    tipsRef.update(_ - key).flatTap(_ => metrics.incrementMetricAsync("checkpointTipsRemoved"))
+
+  def markAsConflict(key: String)(implicit metrics: Metrics): F[Unit] =
+    get(key).flatMap { m =>
+      if (m.isDefined)
+        remove(key)
+          .flatMap(_ => conflictingTips.update(_ + (key -> m.get.checkpointBlock)))
+          .flatTap(_ => Sync[F].delay(logger.warn(s"Marking tip as conflicted tipHash: $key")))
```

Comment:
you can pass logger here as `ConcurrentTipService[F[_]: Concurrent: Logger]` and here you can write:
`.flatTap(_ => Logger[F].warn(s"lorem ipsum"))`